File: 프로그래머스/2/12952. N－Queen/N－Queen.java
class Solution {
    
    public void dfs(int n){
        if(n == N){
            // n개 퀸 배치 완료
            answer++;
            return;
        }
        
        for(int j=0; j<N; j++){
            if(v1[j] == 0 && v2[n+j] == 0 && v3[j-n+(N-1)] == 0){
                v1[j] = 1;
                v2[n+j] = 1;
                v3[j-n+(N-1)] = 1;
                
                dfs(n+1);
                
                v1[j] = 0;
                v2[n+j] = 0;
                v3[j-n+(N-1)] = 0;
            }
        }
    }
    
    static int answer = 0;
    static int N;
    static int[] v1; // 가로 or 세로로 배치된 퀸 체크
    static int[] v2; // 오른쪽 위 배치된 퀸 체크
    static int[] v3; // 왼쪽 위 배치된 퀸 체크
    
    public int solution(int n) {
        N = n;
        v1 = new int[n];
        v2 = new int[n*2-1];
        v3 = new int[n*2-1];
        
        dfs(0);
        
        return answer;
    }
}

File: 프로그래머스/2/12952. N－Queen/N－Queen.py
ans = 0

def solution(n):
    N = n

    v1 = [False]*n # 세로
    v2 = [False]*(2*n) # 왼쪽 대각선
    v3 = [False]*(2*n) # 오른쪽 대각선
    
    def dfs(row):
        global ans
        
        if N == row:
            ''' 
            체스판 마지막 줄에 퀸을 배치할 수 있을 때가 정답 케이스가 된다.
            퀸은 가로, 세로 같은 행과 열에 2개 이상을 배치할 수 없기 때문이다.
            '''
            ans += 1
            return

        for j in range(N):
            # n번째 행에서 0~N번 열을 순회하며 해당 위치에 퀸을 배치할 수 있는지 체크한다.
            if not v1[j] and not v2[row-j] and not v3[row+j]:
                v1[j] = True
                v2[row-j] = True
                v3[row+j] = True

                dfs(row+1) # 다음 행으로 넘어가서 체크

                v1[j] = False
                v2[row-j] = False
                v3[row+j] = False
    
    dfs(0)
    
    return ans